oak_crypto/src/tests.rs

```diff
@@ -0,0 +1,165 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    aead::{AEAD_ALGORITHM_KEY_SIZE_BYTES, AEAD_NONCE_SIZE_BYTES},
+    hpke::{setup_base_recipient, setup_base_sender, KeyPair},
+    util::i2osp,
+    ClientCryptoProvider, EnclaveCryptoProvider,
+};
+
+/// Test AES-GCM key that is only used in tests.
+/// Was generated by calling [`Hpke::setup_base_sender`].
+const TEST_AEAD_KEY: [u8; AEAD_ALGORITHM_KEY_SIZE_BYTES] = [
+    11, 107, 5, 176, 4, 145, 171, 193, 163, 81, 105, 238, 171, 115, 56, 160, 130, 85, 22, 227, 118,
+    76, 77, 89, 144, 223, 10, 112, 11, 149, 205, 199,
+];
+/// Test AES-GCM nonce that is only used in tests.
+const TEST_NONCE: [u8; AEAD_NONCE_SIZE_BYTES] = [
+    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C,
+];
+const TEST_HPKE_INFO: &[u8] = b"Test HPKE info";
+const TEST_REQUEST_MESSAGE: &[u8] = b"Test request message";
+const TEST_REQUEST_ASSOCIATED_DATA: &[u8] = b"Test request associated data";
+const TEST_RESPONSE_MESSAGE: &[u8] = b"Test response message";
+const TEST_RESPONSE_ASSOCIATED_DATA: &[u8] = b"Test response associated data";
+// Number of message exchanges done to test secure session handling.
+const TEST_SESSION_SIZE: usize = 8;
+
+#[test]
+fn test_aead() {
+    let encrypted_message = crate::aead::encrypt(
+        &TEST_AEAD_KEY,
+        &TEST_NONCE,
+        TEST_REQUEST_MESSAGE,
+        TEST_REQUEST_ASSOCIATED_DATA,
+    )
+    .expect("couldn't encrypt test message");
+    // Check that the message was encrypted and the correct associated data was added.
+    assert_ne!(TEST_REQUEST_MESSAGE, encrypted_message.ciphertext);
+    assert_eq!(
+        TEST_REQUEST_ASSOCIATED_DATA,
+        encrypted_message.associated_data
+    );
+    let decrypted_message = crate::aead::decrypt(&TEST_AEAD_KEY, &TEST_NONCE, &encrypted_message)
+        .expect("couldn't decrypt test message");
+    // Test that AEAD decrypts message to the original one.
+    assert_eq!(TEST_REQUEST_MESSAGE, decrypted_message);
+}
+
+#[test]
+fn test_hpke() {
+    let recipient_key_pair = KeyPair::generate();
+    let (serialized_encapsulated_public_key, mut sender_context, mut sender_response_context) =
+        setup_base_sender(
+            &recipient_key_pair.get_serialized_public_key(),
+            TEST_HPKE_INFO,
+        )
+        .expect("couldn't setup base sender");
+    let (mut recipient_context, mut recipient_response_context) = setup_base_recipient(
+        &serialized_encapsulated_public_key,
+        &recipient_key_pair,
+        TEST_HPKE_INFO,
+    )
+    .expect("couldn't setup base recipient");
+
+    // Test that sequence numbers are increased correctly.
```

Comment:
How is this being tested exactly? In the sense that encrypted_request is expected to be different within each iteration?